### PR TITLE
Pm struct optimization

### PR DESCRIPTION
**Pull request by keith-zephyr**
Update the Z_DEVICE_DEFINE macro to conditionally compile in the power management structures.  This uses macro concatenation to detect when the pm_control_fn parameter is NULL.  Due to macro expansion, many of the intermediate macros had to be converted to call Z_DEVICE_DEFINE directly.

This saves considerable RAM space (over 4500 bytes) in a Chrome EC implementation as only a couple drivers actually require the power management callback.

An alternate approach could be to create separate macros for drivers that need PM support, but this would be a substantial API change.

### CODE REVIEW

**Review by jackrosenthal — APPROVED** [review]

File: include/device.h [hunk 1]
```diff
@@ -809,7 +830,7 @@ BUILD_ASSERT(sizeof(device_handle_t) == 2, "fix the linker scripts");
 		.api = (api_ptr),					\
 		.state = &Z_DEVICE_STATE_NAME(dev_name),		\
 		.data = (data_ptr),					\
-		Z_DEVICE_DEFINE_INIT(node_id, dev_name, pm_control_fn)	\
+		Z_DEVICE_DEFINE_INIT(node_id, dev_name, pm_control_fn, use_pm)	\
```
Comment (yperess) [inline include/device.h:833]: nit: formatting

File: include/net/ethernet.h [hunk 2]
```diff
@@ -827,21 +827,21 @@ static inline bool net_eth_get_vlan_status(struct net_if *iface)
 
 #if defined(CONFIG_NET_VLAN)
 #define Z_ETH_NET_DEVICE_INIT(node_id, dev_name, drv_name, init_fn,	\
-			      pm_control_fn, data, cfg, prio, api, mtu)	\
+			      pm_control_fn, data, cfg, prio, api, mtu, use_pm)	\
 	Z_DEVICE_DEFINE(node_id, dev_name, drv_name, init_fn,		\
 			pm_control_fn, data, cfg, POST_KERNEL,		\
-			prio, api);					\
+			prio, api, use_pm);					\
```
Comment (yperess) [inline include/net/ethernet.h:833]: nit: formatting

File: include/net/ethernet.h [hunk 3]
```diff
@@ -827,21 +827,21 @@ static inline bool net_eth_get_vlan_status(struct net_if *iface)
...
 	NET_L2_DATA_INIT(dev_name, 0, NET_L2_GET_CTX_TYPE(ETHERNET_L2));\
 	NET_IF_INIT(dev_name, 0, ETHERNET_L2, mtu, NET_VLAN_MAX_COUNT)
 
 #else /* CONFIG_NET_VLAN */
 
 #define Z_ETH_NET_DEVICE_INIT(node_id, dev_name, drv_name, init_fn,	\
-			      pm_control_fn, data, cfg, prio, api, mtu)	\
+			      pm_control_fn, data, cfg, prio, api, mtu, use_pm)	\
```
Comment (yperess) [inline include/net/ethernet.h:840]: nit: formatting

**Review by yperess — APPROVED** [review]
minor nits, should run clang-format but LGTM

File: include/device.h [hunk 4]
```diff
@@ -783,23 +796,31 @@ BUILD_ASSERT(sizeof(device_handle_t) == 2, "fix the linker scripts");
...
 #define Z_DEVICE_DEFINE_PM_INIT(dev_name, pm_control_fn)
 #endif
 
-#define Z_DEVICE_DEFINE_INIT(node_id, dev_name, pm_control_fn)		\
+#define Z_DEVICE_DEFINE_INIT(node_id, dev_name, pm_control_fn, use_pm)	\
 		.handles = Z_DEVICE_HANDLE_NAME(node_id, dev_name),	\
-		Z_DEVICE_DEFINE_PM_INIT(dev_name, pm_control_fn)
+		COND_CODE_1(use_pm, 					\
+			(Z_DEVICE_DEFINE_PM_INIT(dev_name, pm_control_fn)), \
+			())
```
Comment (yperess) [inline include/device.h:813]: I think that these lines are causing the issue, I think that if `use_mp` is false you need to set `pm_control = NULL` explicitly (might be wrong).

**Review by gmarull — COMMENTED** [review]
Thanks for these changes, +1 to get this feature in.

I tried to achieve something similar in the past here #35257, but removing usage of `NULL`. The fact that we use `NULL` means we can't freely use it as a pre-processor token, since it gets expanded (`(void *)0`). Another option would be to stop using `NULL` if no PM is present and use a token, e.g. `PM_DISABLE`. This way, I assume most of the init macros would be left untouched and `use_pm` would not be required. Opinions?

Note: Adding DNM to avoid merging without discussion

**Review by gmarull — CHANGES_REQUESTED** [review]
So adding -1 to avoid merging without discussion first, not to block.

**Comment by gmarull** [comment]
> 
> I also considered creating a new token, such as `PM_DISABLED`. Would this be considered an API change, because the `pm_control_fn` goes from specifying `NULL` to `PM_DISABLED`? The only downside I see to a new token is that users are probably more accustomed to passing in NULL for an used pointer.

I'm not sure if it is considered an API change, but it certainly is an annoying change. Another option to consider could be to decouple PM from `DEVICE_DEFINE` and use a separate macro, same as it has been done for the pinctrl proposal, see [this](https://github.com/zephyrproject-rtos/zephyr/blob/14a560fe01491a150123f76f6c1203d9afb76394/doc/guides/pinctrl/index.rst#device-drivers). Device PM fields could then be initialized using _special names_ crafted using `DEVICE_(DT)_NAME_GET` (see [this](https://github.com/zephyrproject-rtos/zephyr/blob/93d34d80aa80d6db8d01fec7baddab0dea220b63/include/drivers/pinctrl.h#L93)). A separate macro would allow future extensions without modifying device related macros (e.g. add some more fields to indicate supported features or required functionality such as async runtime). Just an idea, should be seen if feasible/reasonable.

> 
> I'll submit a separate PR that uses the `PM_DISABLED` approach. I think this can be done as 2 PRs. 1st PR adds `#define PM_DISABLED NULL` and changes all existing references from `NULL` to `PM_DISABLED`. 2nd PR adds in the logic to compile out the PM structs when `PM_DISABLED` detected.

A quick proof of concept with a board that compiles should be enough :-)

**Review by teburd — CHANGES_REQUESTED** [review]
This change is really quite nice when it comes to memory usage, a quick test on frdm k64 shows significant memory savings.

I'm torn with the idea that we need a new token rather than NULL. I like that the device macro API is almost entirely unchanged with the null detection bits, not breaking other peoples code is always a nice attribute of a change.

**Comment by ceolin** [comment]
Nice change !  Have to carry on the `use_pm` condition for all sub-macros because you can only concatenate the null parameter in the top level macro is really cumbersome and hard to read, but that is the life with C preprocessor. I think this is more natural for an API than have to use a specific token. +1 for this approach

**Comment by keith-zephyr** [comment]
[PR#39397](https://github.com/zephyrproject-rtos/zephyr/pull/39397) merged and addresses this issue.

**Pull request closed without merging**